pkg/servers/ipfix/ipfix_server.go

```diff
@@ -0,0 +1,329 @@
+	"io"
+	"net"
+	"runtime/debug"
+	"strconv"
+	"strings"
+	"sync"
+
+	bnet "github.com/bio-routing/bio-rd/net"
+	"github.com/bio-routing/flowhouse/pkg/models/flow"
+	"github.com/bio-routing/flowhouse/pkg/packet/ipfix"
+	"github.com/bio-routing/tflow2/convert"
+	"github.com/pkg/errors"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type InterfaceResolver interface {
+	Resolve(agent bnet.IP, ifID uint32) string
+}
+
+// fieldMap describes what information is at what index in the slice
+// that we get from decoding a netflow packet
+type fieldMap struct {
+	srcAddr                int
+	dstAddr                int
+	protocol               int
+	packets                int
+	size                   int
+	intIn                  int
+	intOut                 int
+	nextHop                int
+	family                 int
+	vlan                   int
+	ts                     int
+	srcAsn                 int
+	dstAsn                 int
+	srcPort                int
+	dstPort                int
+	samplingPacketInterval int
+}
+
+type IPFIXServer struct {
+	// tmplCache is used to save received flow templates
+	// for later lookup in order to decode netflow packets
+	tmplCache  *templateCache
+	conn       *net.UDPConn
+	ifResolver InterfaceResolver
+	output     chan []*flow.Flow
+	wg         sync.WaitGroup
+	stopCh     chan struct{}
+}
+
+// New creates and starts a new `IPFIXServer` instance
+func New(listen string, numReaders int, output chan []*flow.Flow, ifResolver InterfaceResolver) (*IPFIXServer, error) {
+	ipf := &IPFIXServer{
+		tmplCache:  newTemplateCache(),
+		ifResolver: ifResolver,
+		stopCh:     make(chan struct{}),
+		output:     output,
+	}
+
+	addr, err := net.ResolveUDPAddr("udp", listen)
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to resolve UDP address")
+	}
+
+	con, err := net.ListenUDP("udp", addr)
+	if err != nil {
+		return nil, errors.Wrap(err, "ListenUDP failed")
+	}
+	ipf.conn = con
+
+	ipf.startService(numReaders)
+	return ipf, nil
+}
+
+func (ipf *IPFIXServer) startService(numReaders int) {
+	for i := 0; i < numReaders; i++ {
+		ipf.wg.Add(1)
+		go func() {
+			defer ipf.wg.Done()
+			err := ipf.packetWorker()
+			if err != nil {
+				log.WithError(err).Error("packetWorker failed")
+			}
+		}()
+	}
+}
+
+// Stop closes the socket and stops the workers
+func (ipf *IPFIXServer) Stop() {
+	log.Info("Stopping IPFIX server")
+	debug.PrintStack()
+	close(ipf.stopCh)
+	ipf.conn.Close()
+	ipf.wg.Wait()
+}
+
+// packetWorker reads sflow packet from socket and handsoff processing to ???
+func (ipf *IPFIXServer) packetWorker() error {
+	buffer := make([]byte, 8960)
+	for {
+		if ipf.stopped() {
+			return nil
+		}
+
+		length, remote, err := ipf.conn.ReadFromUDP(buffer)
+		if err == io.EOF {
+			return nil
+		}
+
+		if err != nil {
+			return errors.Wrap(err, "ReadFromUDP failed")
+		}
+
+		remote4 := remote.IP.To4()
+		if remote4 != nil {
+			remote.IP = remote4
+		}
+
+		remoteAddr, err := bnet.IPFromBytes([]byte(remote.IP))
```

Comment:
If remote.IP is a IPv6, this will still work?

Reply:
I don't see a reason it would not. bnet.IPFromBytes supports both IPv4 and IPv6.